enhancements/authentication/improved-login-workflow.md

```diff
@@ -0,0 +1,255 @@
+implementation and anything particularly challenging to test should be called out.
+
+All code is expected to have adequate tests (eventually with coverage expectations).
+
+### Graduation Criteria
+
+**Note:** *Section not required until targeted at a release.*
+
+Define graduation milestones.
+
+These may be defined in terms of API maturity, or as something else. Initial proposal should keep this high-level with a
+focus on what signals will be looked at to determine graduation.
+
+Consider the following in developing the graduation criteria for this enhancement:
+
+- Maturity levels
+  - [`alpha`, `beta`, `stable` in upstream Kubernetes][maturity-levels]
+  - `Dev Preview`, `Tech Preview`, `GA` in OpenShift
+- [Deprecation policy][deprecation-policy]
+
+Clearly define what graduation means by either linking to
+the [API doc definition](https://kubernetes.io/docs/concepts/overview/kubernetes-api/#api-versioning), or by redefining
+what graduation means.
+
+In general, we try to use the same stages (alpha, beta, GA), regardless how the functionality is accessed.
+
+[maturity-levels]: https://git.k8s.io/community/contributors/devel/sig-architecture/api_changes.md#alpha-beta-and-stable-versions
+
+[deprecation-policy]: https://kubernetes.io/docs/reference/using-api/deprecation-policy/
+
+**Examples**: These are generalized examples to consider, in addition to the
+aforementioned [maturity levels][maturity-levels].
+
+#### Dev Preview -> Tech Preview
+
+- Ability to utilize the enhancement end to end
+- End user documentation, relative API stability
+- Sufficient test coverage
+- Gather feedback from users rather than just developers
+- Enumerate service level indicators (SLIs), expose SLIs as metrics
+- Write symptoms-based alerts for the component(s)
+
+#### Tech Preview -> GA
+
+- More testing (upgrade, downgrade, scale)
+- Sufficient time for feedback
+- Available by default
+- Backhaul SLI telemetry
+- Document SLOs for the component
+- Conduct load testing
+
+**For non-optional features moving to GA, the graduation criteria must include end to end tests.**
+
+#### Removing a deprecated feature
+
+NA
+
+### Upgrade / Downgrade Strategy
+
+NA
+
+### Version Skew Strategy
+
+#### Compatibility of new oc login flow with previous versions of OCP
+
+When the new login flow is used `oc` will open a new browser window which will make a request on the `/authorize`
+endpoint. The client-id needs to be passed here, and it will be the name of the new `OAuthClient` resource created by
+the `cluster-authentication-operator`.
+
+_Newer oc with older API server_: In older versions of OCP where the `cluster-authentication-operator` has not created
+this new client the request to the OAuth server will fail because the client id is invalid. The redirect to the endpoint
+on localhost will not happen and an error message will be displayed in the browser window.
```

Comment:
Could we detect what we're allowed to do by using something like https://datatracker.ietf.org/doc/html/rfc8414, which is somewhat built in to the kube-apiserver even today?

Reply:
The metadata endpoint would not expose which clients are registered and what callback URLs they have, because this would be confidential. So it's not possible to determine if the the new worklflow can be used from just the metadata itself.